stomp_plugins/src/cost_functions/cartesian_distance.cpp

```diff
@@ -0,0 +1,182 @@
+#include <math.h>
+#include <stomp_plugins/cost_functions/cartesian_distance.h>
+#include <XmlRpcException.h>
+#include <pluginlib/class_list_macros.h>
+#include <ros/console.h>
+
+PLUGINLIB_EXPORT_CLASS(stomp_moveit::cost_functions::CartesianDistance,stomp_moveit::cost_functions::StompCostFunction);
+
+namespace stomp_moveit
+{
+namespace cost_functions
+{
+
+CartesianDistance::CartesianDistance():
+    name_("CartesianDistance")
+{
+  // TODO Auto-generated constructor stub
+
+}
+
+CartesianDistance::~CartesianDistance()
+{
+  // TODO Auto-generated destructor stub
+}
+
+bool CartesianDistance::initialize(moveit::core::RobotModelConstPtr robot_model_ptr,
+                        const std::string& group_name,XmlRpc::XmlRpcValue& config)
+{
+  group_name_ = group_name;
+  robot_model_ = robot_model_ptr;
+
+  return configure(config);
+}
+
+bool CartesianDistance::configure(const XmlRpc::XmlRpcValue& config)
+{
+  using namespace XmlRpc;
+
+  try
+  {
+    XmlRpcValue params = config;
+
+    position_cost_weight_ = static_cast<double>(params["position_cost_weight"]);
+    orientation_cost_weight_ = static_cast<double>(params["orientation_cost_weight"]);
+    translation_tolerance_ =  static_cast<double>(params["translation_tolerance"]);
+    rotation_tolerance_ =  static_cast<double>(params["rotation_tolerance"]);
+
+    if (translation_tolerance_ <= 0)
+    {
+      ROS_ERROR("%s the 'translation_tolerance' parameter must not have non positive value",getName().c_str());
+      return false;
+    }
+
+    if (rotation_tolerance_ <= 0)
+    {
+      ROS_ERROR("%s the 'rotation_tolerance_' parameter must not have non positive value",getName().c_str());
+      return false;
+    }
+
+    // total weight
+    cost_weight_ = position_cost_weight_ + orientation_cost_weight_;
+
+  }
+  catch(XmlRpc::XmlRpcException& e)
+  {
+    ROS_ERROR("%s failed to load parameters, %s",getName().c_str(),e.getMessage().c_str());
+    return false;
+  }
+
+  return true;
+}
+
+bool CartesianDistance::setMotionPlanRequest(const planning_scene::PlanningSceneConstPtr& planning_scene,
+                 const moveit_msgs::MotionPlanRequest &req,
+                 const stomp_core::StompConfiguration &config,
+                 moveit_msgs::MoveItErrorCodes& error_code)
+{
+  using namespace moveit::core;
+
+  const JointModelGroup* joint_group = robot_model_->getJointModelGroup(group_name_);
+  int num_joints = joint_group->getActiveJointModels().size();
+  tool_link_ = joint_group->getLinkModelNames().back();
```

Comment:
Class member variable should be initialized in constructor. Declare id locally or initialize in constructor. Are you sure that tool link is always on the end of chain? Should the tool link name be in config?

Reply:
I add tool link name to config